worlds/pokemon_emerald/rules.py

```diff
@@ -0,0 +1,1345 @@
+"""
+Logic rule definitions for Pokemon Emerald
+"""
+from BaseClasses import CollectionState, MultiWorld
+
+from worlds.generic.Rules import add_rule, set_rule
+
+from .options import EliteFourRequirement, NormanRequirement, ExtraBoulders
+from .util import location_name_to_label
+
+
+def _can_cut(state: CollectionState, player: int):
+    return state.has("HM01 Cut", player) and state.has("Stone Badge", player)
+def _can_fly(state: CollectionState, player: int):
+    return state.has("HM02 Fly", player) and state.has("Feather Badge", player)
+def _can_surf(state: CollectionState, player: int):
+    return state.has("HM03 Surf", player) and state.has("Balance Badge", player)
+def _can_strength(state: CollectionState, player: int):
+    return state.has("HM04 Strength", player) and state.has("Heat Badge", player)
+def _can_flash(state: CollectionState, player: int):
+    return state.has("HM05 Flash", player) and state.has("Knuckle Badge", player)
+def _can_rock_smash(state: CollectionState, player: int):
+    return state.has("HM06 Rock Smash", player) and state.has("Dynamo Badge", player)
+def _can_waterfall(state: CollectionState, player: int):
+    return state.has("HM07 Waterfall", player) and state.has("Rain Badge", player)
+def _can_dive(state: CollectionState, player: int):
+    return state.has("HM08 Dive", player) and state.has("Mind Badge", player)
+
+def _can_use_mach_bike(state: CollectionState, player: int):
+    return state.has("Mach Bike", player)
+def _can_use_acro_bike(state: CollectionState, player: int):
+    return state.has("Acro Bike", player)
+
+def _defeated_n_gym_leaders(state: CollectionState, player: int, n: int):
+    num_gym_leaders_defeated = 0
+    num_gym_leaders_defeated += 1 if state.has("EVENT_DEFEAT_ROXANNE", player) else 0
+    num_gym_leaders_defeated += 1 if state.has("EVENT_DEFEAT_BRAWLY", player) else 0
+    num_gym_leaders_defeated += 1 if state.has("EVENT_DEFEAT_WATTSON", player) else 0
+    num_gym_leaders_defeated += 1 if state.has("EVENT_DEFEAT_FLANNERY", player) else 0
+    num_gym_leaders_defeated += 1 if state.has("EVENT_DEFEAT_NORMAN", player) else 0
+    num_gym_leaders_defeated += 1 if state.has("EVENT_DEFEAT_WINONA", player) else 0
+    num_gym_leaders_defeated += 1 if state.has("EVENT_DEFEAT_TATE_AND_LIZA", player) else 0
+    num_gym_leaders_defeated += 1 if state.has("EVENT_DEFEAT_JUAN", player) else 0
+    return num_gym_leaders_defeated >= n
+
+def _has_at_least_n_badges(state: CollectionState, player: int, n: int):
```

Comment:
There is a function `has_group` that you should use instead of this.

Reply:
Changed in 4a382da

I assume you didn't comment on the defeated gym leaders version because it doesn't make sense to create an item group for events?